src/runtime_src/core/common/drv/kds_core.c

```diff
@@ -1271,12 +1266,25 @@ void kds_fini_client(struct kds_sched *kds, struct kds_client *client)
 	struct kds_client_hw_ctx *curr = NULL;
 	struct kds_client_ctx *c_curr = NULL;
 
-	/* Release legacy client's resources */
-	_kds_fini_client(kds, client, client->ctx);
+	/* Release legacy client's resources for XOCL */
+	if (client->ctx) {
+		_kds_fini_client(kds, client, client->ctx);
+		mutex_lock(&client->lock);
+		curr = kds_get_hw_ctx_by_id(client, 0 /* default hw cx id */);
+		if (curr)
+			kds_free_hw_ctx(client, curr);
+		mutex_unlock(&client->lock);
+	}
+	/* this is for ZOCL legacy client */
```

Comment:
I don't feel very comfortable about this comment. This is the shared code. It should not have platform specific code. Do we have a plan to remove the legacy client?

Reply:
Fixed. Please check

Reply:
Look good to me. Thanks